drivers/video/video_st_mipid02.c

```diff
@@ -0,0 +1,456 @@
+		if (desc) {
+			LOG_ERR("Sensor format not supported by the ST-MIPID02");
+			return -EIO;
+		}
+
+		/* Override the pixelformat */
+		fmt->pixelformat = desc->pixelformat;
+
+		drv_data->fmt = *fmt;
+		drv_data->desc = desc;
+	}
+
+	*fmt = drv_data->fmt;
+
+	return 0;
+}
+
+static int mipid02_get_caps(const struct device *dev, struct video_caps *caps)
+{
+	const struct mipid02_config *cfg = dev->config;
+	struct mipid02_data *drv_data = dev->data;
+	int ind = 0, caps_nb = 0;
+	int ret;
+
+	ret = video_get_caps(cfg->source_dev, caps);
+	if (ret < 0) {
+		return ret;
+	}
+
+	while (caps->format_caps[ind].pixelformat) {
+		if (mipid02_get_format_desc(caps->format_caps[ind].pixelformat)) {
+			caps_nb++;
+		}
+		ind++;
+	}
+
+	k_heap_free(&mipid02_video_caps_heap, drv_data->caps);
+	drv_data->caps = k_heap_alloc(&mipid02_video_caps_heap,
+				      (caps_nb + 1) * sizeof(struct video_format_cap), K_FOREVER);
+	if (!drv_data->caps) {
+		return -ENOMEM;
+	}
+
+	ind = 0;
+	for (int i = 0; caps->format_caps[i].pixelformat; i++) {
+		const struct mipid02_format_desc *desc =
+			mipid02_get_format_desc(caps->format_caps[i].pixelformat);
+
+		if (!desc) {
+			continue;
+		}
+
+		drv_data->caps[ind] = caps->format_caps[i];
+		drv_data->caps[ind].pixelformat = desc->pixelformat;
+		ind++;
+
+		if (ind == caps_nb) {
+			memset(&drv_data->caps[ind], 0, sizeof(drv_data->caps[ind]));
+			break;
+		}
+	}
+
+	caps->format_caps = drv_data->caps;
+	caps->min_vbuf_count = 1;
+	caps->min_line_count = LINE_COUNT_HEIGHT;
+	caps->max_line_count = LINE_COUNT_HEIGHT;
+
+	return 0;
+}
+
+static int mipid02_set_stream(const struct device *dev, bool enable, enum video_buf_type type)
+{
+	const struct mipid02_config *cfg = dev->config;
+	struct mipid02_data *drv_data = dev->data;
+	const struct mipid02_format_desc *desc;
+	int32_t link_freq;
+	int ret;
+
+	if (!enable) {
+		ret = video_stream_stop(cfg->source_dev, type);
+		if (ret < 0) {
+			return ret;
+		}
+
+		/* Disable lanes */
+		ret = video_write_cci_reg(&cfg->i2c, MIPID02_CLK_LANE_REG1, 0);
+		if (ret < 0) {
+			return ret;
+		}
+
+		ret = video_write_cci_reg(&cfg->i2c, MIPID02_DATA_LANE0_REG1, 0);
+		if (ret < 0) {
+			return ret;
+		}
+
+		if (cfg->csi.nb_lanes == 2) {
+			ret = video_write_cci_reg(&cfg->i2c, MIPID02_DATA_LANE1_REG1, 0);
+			if (ret < 0) {
+				return ret;
+			}
+		}
+	}
+
+	desc = mipid02_get_format_desc(drv_data->fmt.pixelformat);
+	if (!desc) {
+		LOG_ERR("No valid format desc available, should get/set_fmt prior to set_stream");
+		return -EIO;
+	}
+
+	/* Get link-frequency information from source */
+	link_freq = video_get_link_frequency(cfg->source_dev,
+					     video_bits_per_pixel(desc->csi_pixelformat),
+					     cfg->csi.nb_lanes);
+	if (link_freq < 0) {
+		LOG_ERR("Failed to retrieve source link-frequency");
+		return -EIO;
+	}
+
+	/* Enable lanes */
+	ret = video_write_cci_reg(&cfg->i2c, MIPID02_CLK_LANE_REG1,
+				  (2000000000 / link_freq) << MIPID02_CLK_LANE_REG1_UI_SHIFT |
```

Comment:
[non-blocking]
If you wanted to try [`FIELD_PREP`](https://docs.zephyrproject.org/latest/doxygen/html/group__sys-util.html#gaa03c8b31bf67a097dd9f8153a04ef86b):
```suggestion
				  FIELD_PREP(MIPID02_CLK_LANE_REG1_UI_MASK, 2000000000 / link_freq) |
```